csrc/multidevice/communication.cpp

```diff
@@ -40,56 +42,148 @@ inline void assertBuffersHaveSameSize(
   }
 }
 
-inline void post_common(Communication& self, Communicator& comm) {
-  NVF_ERROR(
-      std::find(
-          self.params().team.begin(),
-          self.params().team.end(),
-          comm.deviceId()) != self.params().team.end(),
-      "current device index ",
-      comm.deviceId(),
-      " must be present in the communication's team");
-}
-
 inline void doLocalCopy(const at::Tensor& dst, const at::Tensor& src) {
   dst.copy_(src, /* non-blocking */ true);
 }
 
-} // namespace
+inline bool hasRoot(CommunicationType type) {
+  return type == CommunicationType::Gather ||
+      type == CommunicationType::Scatter ||
+      type == CommunicationType::Broadcast ||
+      type == CommunicationType::SendRecv;
+}
+
+inline bool isReduction(CommunicationType type) {
+  return type == CommunicationType::Reduce ||
+      type == CommunicationType::Allreduce ||
+      type == CommunicationType::ReduceScatter;
+}
+
+inline std::string typeToString(CommunicationType type) {
+  switch (type) {
+    case CommunicationType::Gather:
+      return "Gather";
+    case CommunicationType::Allgather:
+      return "Allgather";
+    case CommunicationType::Scatter:
+      return "Scatter";
+    case CommunicationType::Reduce:
+      return "Reduce";
+    case CommunicationType::Allreduce:
+      return "Allreduce";
+    case CommunicationType::ReduceScatter:
+      return "ReduceScatter";
+    case CommunicationType::Broadcast:
+      return "Broadcast";
+    case CommunicationType::SendRecv:
+      return "SendRecv";
+    default:
+      NVF_ERROR(false);
+      return "";
+  }
+}
 
-Communication::Communication(CommParams params, std::string name, bool has_root)
-    : params_(std::move(params)),
-      collective_type_(std::move(name)),
-      has_root_(has_root) {
-  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
+inline void assertValid(
+    const CommParams& params,
+    const DeviceIdxType my_device_index) {
+  assertBuffersHaveSameSize(params.src_bufs, params.dst_bufs);
   NVF_ERROR(
-      std::unique(params_.team.begin(), params_.team.end()) ==
-          params_.team.end(),
+      std::adjacent_find(params.team.cbegin(), params.team.cend()) ==
```

Comment:
Question: this assumes params.team is sorted. Is that guaranteed?

Reply:
that's a mistake on my side, let me fix that

Reply:
I reverted to `std::unique` but notice that I need to copy the team because `unique` doesn't preserve const-ness. Let me know if you prefer another implementation. The goal is simply to check that there are no duplicates